nathanielmillard/rancid-tomatillos
# Nd initial testing

#### What’s this PR do?
- This PR adds unit testing for each individual component that has been built so far.
- Fixes a bug reference where `class` was used instead of `className`.
- Installed these libraries to get testing to work: `testing-library/react`, `testing-library/user-event`, `testing-library/jest-dom`.
- Updated component file structure to separate each individual component into it's own folder with associated test file.
- Made adjustments to import Link component from react-router-dom into NavBar component to prevent the error `Link should not be used outside of a Router`.

#### Where should the reviewer start?
- Git pull this branch.

#### How should this be manually tested?
- Run `npm install` to install the new dev dependency libraries mentioned above.
- Run `npm start` to open server on localhost:3000.
- Navigate to localhost:3000 and click on Login button to ensure it still functions properly.
- Sign in with a user with the credentials `email: lucy@turing.io, password: password1` to ensure you are redirected back to main page after successful sign in.
- Try signing in with wrong credentials to get an alert error message.
- Run `npm test` inside terminal while inside this repo directory to see test file assertions.
- Ensure all tests are passing.

#### Any background context you want to provide?
- There are many tests that are commented out as they are not possible to finish without async testing. Since this just got the testing suite started and we have not fully completed the single Movie page yet, will research async testing later or after lesson to implement.
- There are multiple nuisances to the testing, such as having an alert inside of NavBar.js, that were throwing many error messages. `global.alert = jest.fn()` fixed this error message.
- Without async testing, many tests are commented out and only focused on mainly unit testing rather than integration as async will be needed for that.

#### What are the relevant tickets?
- Closes #11
- Closes #12
- While navigation has not been fully testing yet, Those tickets will not be added here as closed yet.

#### Screenshots (if appropriate):
- N/A

#### Questions:
- Interesting note. I noticed while reviewing the files that testing-library/jest-dom, testing-library/user-event, testing-library react were already part of the package.json, just not as a dev dependency and a much lower version number. Does that mean `create-react-app` pre-installs with all those as well?
